util/git/client.go

```diff
@@ -322,15 +325,39 @@ func (m *nativeGitClient) Init() error {
 	if err != nil {
 		return err
 	}
-	repo, err := git.PlainInit(m.root, false)
+
+	_, err = m.runCmd("init")
 	if err != nil {
 		return err
 	}
-	_, err = repo.CreateRemote(&config.RemoteConfig{
-		Name: git.DefaultRemoteName,
-		URLs: []string{m.repoURL},
-	})
-	return err
+
+	_, err = m.runCmd("remote", "add", git.DefaultRemoteName, m.repoURL)
+	if err != nil {
+		return err
+	}
+	return nil
+}
+
+// SparseCheckout initializes a git sparse-checkout, and sets the patterns for the sparse-checkout.
+func (m *nativeGitClient) SparseCheckout(patterns []string) error {
+	if _, err := m.runCmd("sparse-checkout", "init"); err != nil {
+		return err
+	}
+
+	var sb strings.Builder
+
+	for i, pattern := range patterns {
+		if i > 0 {
+			sb.WriteString(" ")
+		}
+		sb.WriteString(pattern)
+	}
+
+	if _, err := m.runCmd("sparse-checkout", "set", sb.String()); err != nil {
```

Comment:
If I understand this correctly, the sparse-checkout setting "sticks." I think this means we need to prevent concurrent access to the repo for apps where the sparse checkout setting doesn't match (similar to what we do for Kustomize apps where some on-disk state is set by a config option).

Reply:
It does stick to the configuration of a git repo. Gonna be working on that next.

Reply:
> similar to what we do for Kustomize apps where some on-disk state is set by a config option

@crenshaw-dev I am gonna be digging into it, but in the meantime, could you share some links to where the Kustomize version is happening?

Trying to figure out how the codebase reasons about such a situation.

Reply:
There's a semaphore mechanism to allow multiple read-only processes to use a repo checkout out at a particular commit at the same time. 

I think this is the field which determines whether concurrency is allowed for a certain repo operation: https://github.com/argoproj/argo-cd/blob/c200d0c7646ed6b327d6db686f456c24cf095aeb/reposerver/repository/repository.go#L259